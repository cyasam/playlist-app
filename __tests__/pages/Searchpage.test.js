import React from 'react'
import { shallow } from 'enzyme'
import { Searchpage, mapStateToProps } from '../../src/scripts/pages/Searchpage'

describe('SearchPage Component', () => {
  const setup = (propOverrides, stateOverrides) => {
    const state = {
      search: {
        isFetching: false,
        videos: [
            {
              id: 'lZoA5ZX4wC0',
              title: 'Video Title',
              description: 'Lorem ipsum dolor sit amed.',
              thubmnail: 'video-thumbnail.jpg'
            },
            
            {
              id: 'lZoA5ZX4wC01',
              title: 'Video Title 2',
              description: 'Lorem ipsum dolor sit amed.',
              thubmnail: 'video-thumbnail-2.jpg'
            }
        ],
        query: 'abc',
        nextPageToken: 'CBgQAA'
      },
      ...stateOverrides
    }

    const props = {
      searchResult: state.search,
      fetchSearch: jest.fn(),
      loadmoreSearch: jest.fn(),
      history: { 
        location: { search: '?query=serdar' },
      },
      ...propOverrides
    }
    const wrapper = shallow(<Searchpage { ...props } />)
    return {
      wrapper,
      props,
      state
    }
  }
  
  it('renders properly', () => {
    const { wrapper } = setup()
    expect(wrapper).toMatchSnapshot()
  })

  it('calls mountPage method when first page load', () => {
    const { wrapper } = setup(null, { search: { isFetching: true, videos: [] } })
    const mockMountPage = jest.spyOn(wrapper.instance(), 'mountPage')

    wrapper.instance().componentDidMount()
    expect(mockMountPage.mock.calls.length).toEqual(1)
  })

  it('pushs history to Homepage action when search query id is not defined in query string', () => {
    const { wrapper, props } = setup({ 
      history: {
        push: jest.fn(),
        location: {
          search: '?query='
        }
      }
    }, { search: { isFetching: true, videos: [] } })

    wrapper.instance().componentDidMount()
    expect(props.history.push).toHaveBeenCalled()
  })

  it('renders properly when search query is not exist', () => {
    const { wrapper, props } = setup({
      history: { 
        location: { search: null }
      },
    })
    expect(props.fetchSearch()).toMatchSnapshot()
  })

  it('returns searchResult object when initialize component', () => {
    const { wrapper, state } = setup()
    expect(mapStateToProps(state).searchResult).toEqual(state.search)
  })

  it('calls loadmoreSearch when calling loadMore', () => {
    const { wrapper, props, state } = setup()
    wrapper.instance().loadMore('abc')
    expect(props.loadmoreSearch).toHaveBeenCalledWith(state.search.query, 'abc')
  })
})