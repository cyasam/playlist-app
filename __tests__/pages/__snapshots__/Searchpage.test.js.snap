// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchPage Component renders Loading when isFetching is true 1`] = `<Loading />`;

exports[`SearchPage Component renders properly 1`] = `
<React.Fragment>
  <h3
    className="main-title"
  >
    Search Results
  </h3>
  <VideoList
    loadMoreVideos={[MockFunction]}
  />
</React.Fragment>
`;

exports[`SearchPage Component renders properly when search query is not exist 1`] = `undefined`;
